pkg/certificate/x509.go

```diff
@@ -0,0 +1,147 @@
+package certificate
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+)
+
+const (
+	rootName    = "rootCA.pem"
+	rootKeyName = "rootCA-key.pem"
+)
+
+func GenerateSignedCertificate(serverName string, priv *rsa.PrivateKey) ([]byte, *rsa.PrivateKey, error) {
+	if len(serverName) == 0 {
+		return nil, nil, errors.New("while generating certificate: server name is empty")
+	}
+
+	caPath, err := mkcertCARootPath()
+	if err != nil {
+		return nil, nil, fmt.Errorf("while generating certificate: %w", err)
+	}
+
+	caCert, err := LoadCertificate(caPath + "/" + rootName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while loading CA certificate: %w", err)
+	}
+
+	caPrivateKey, err := LoadPrivateKey(caPath + "/" + rootKeyName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while loading CA key: %w", err)
+	}
+
+	//nolint:exhaustruct
+	tpl := &x509.Certificate{
+		SerialNumber: &big.Int{}, // mandatory, but useless as it should be a unique id given to the certificate by the CA
+		Subject: pkix.Name{ // not necessary, but cool to have if the user ask for certificate details.
+			CommonName:   serverName,
+			Organization: []string{"thyra dynamically generated"},
+		},
+		NotBefore: time.Now(),
+		// one day of validity is enough since the certificate is generated dynamically each time
+		NotAfter: time.Now().AddDate(0, 0, 1),
+	}
+
+	tpl.DNSNames = append(tpl.DNSNames, serverName)
+
+	// Create the certificate.
+	cert, err := x509.CreateCertificate(rand.Reader, tpl, caCert, priv.Public(), caPrivateKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while creating certificate: %w", err)
+	}
+
+	return cert, priv, nil
+}
+
+// Get the mkcert CA root path depending on the OS.
+func mkcertCARootPath() (string, error) {
+	if env := os.Getenv("CAROOT"); env != "" {
+		return env, nil
+	}
+
+	var dir string
+
+	switch {
+	case runtime.GOOS == "windows":
+		dir = os.Getenv("LocalAppData")
```

Comment:
How would this behave on a computer setup in Chinese/Indian/Arabic ? Does this environment variable exists on all Windows distributions ?

Reply:
I think so but I'm not sure (windows :shrug:).  In anycase, this is exactly what mkcert is doing and as mkcert is the current source of the CA...

Reply:
Fair enough